### PR TITLE
Serialize shared functions and share them between environments

### PR DESCRIPTION
**Pull request by balderdash**
This saves function definitions in their full xml form under `dashboard/config/shared_functions/<level type>/<name>.xml`, and otherwise is basically the same as block serializing/seeding.

### CODE REVIEW

File: dashboard/app/controllers/shared_blockly_functions_controller.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class SharedBlocklyFunctionsController < ApplicationController
   load_and_authorize_resource
 
   def new
+    @shared_blockly_function.block_type = 'behavior'
```
Comment (joshlory) [inline dashboard/app/controllers/shared_blockly_functions_controller.rb:7]: I think this is covered by the `before_save` in the model?

https://github.com/code-dot-org/code-dot-org/blob/1c67af3e880d9568c7fb52a68714012091bfe8e1/dashboard/app/models/shared_blockly_function.rb#L24-L27

Comment (joshlory) [reply dashboard/app/controllers/shared_blockly_functions_controller.rb:7]: Ah I see — removed below.

File: dashboard/app/models/shared_blockly_function.rb [hunk 2]
```diff
@@ -12,17 +12,94 @@
 #  stack       :text(65535)
 #
 
+DEFINITION_BLOCK_TYPES = {
+  function: 'procedures_defnoreturn',
+  behavior: 'behavior_definition',
+}
+
+BLOCK_TYPES_BY_DEFINITION_TYPE = DEFINITION_BLOCK_TYPES.invert
```
Comment (joshlory) [inline dashboard/app/models/shared_blockly_function.rb:20]: 👍

File: dashboard/app/models/shared_blockly_function.rb [hunk 3]
```diff
@@ -12,17 +12,94 @@
...
+    path = xml_path(old_level_type, old_name)
+    File.delete path if File.exist? path
+  end
+
+  def self.load_functions
+    function_names = []
+    LevelLoader.for_each_file('config/shared_functions/**/*.xml') do |xml_path|
+      function_names << load_function(xml_path)
+    end
+    SharedBlocklyFunction.where.not(name: function_names).destroy_all
```
Comment (joshlory) [inline dashboard/app/models/shared_blockly_function.rb]: Is there a race condition here, where a new record can be saved to the DB but not yet written to a file (or written after the glob completes)?

File: dashboard/app/models/shared_blockly_function.rb [hunk 4]
```diff
@@ -12,17 +12,94 @@
...
+    name = function_doc.xpath('/block/title[@name="NAME"]/text()').text
+    stack = function_doc.xpath('/block/statement[@name="STACK"]/*')
+
+    function = find_or_initialize_by(name: name)
+    function.level_type = level_type
+    function.block_type = BLOCK_TYPES_BY_DEFINITION_TYPE[block_type]
+    function.arguments = arguments
+    function.description = description
+    function.stack = stack
+    function.save!
```
Comment (joshlory) [inline dashboard/app/models/shared_blockly_function.rb]: Would love to see this added code de-duped against the similar functionality in https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/block.rb.

File: dashboard/app/models/shared_blockly_function.rb [hunk 5]
```diff
@@ -12,17 +12,94 @@
...
+  def self.arguments_from_xml(args_xml)
+    JSON.generate(
+      args_xml.map do |arg_xml|
+        [arg_xml.attribute('name'), arg_xml.attribute('type')]
+      end.to_h
+    )
+  end
+
+  def xml_path(type=level_type, function_name=name)
+    Rails.root.join "config/shared_functions/#{type}/#{function_name}.xml"
```
Comment (joshlory) [inline dashboard/app/models/shared_blockly_function.rb]: For blocks (where name must be globally unique) I think using `name` makes sense as the filename.  For functions, does it make sense to rely on something other than name?

`name` may still be the best option, just want to pose the question.

File: dashboard/app/models/shared_blockly_function.rb (hunk 3 above)
Comment (balderdash) [reply dashboard/app/models/shared_blockly_function.rb]: Yeah, if you save a record while seeding is happening, the file will not have been found by the initial `Dir.glob` call, so it won't get into `function_names` and thus get deleted. I updated to write the file before saving, and to only remove records that were present before I grab the list of files.

File: dashboard/app/models/shared_blockly_function.rb (hunk 5 above)
Comment (balderdash) [reply dashboard/app/models/shared_blockly_function.rb]: I think it's still a good idea to enforce name uniqueness, even if we later decide to reference them some other way. I can't think of a situation where we really need two functions or behaviors with the same name.

File: dashboard/app/models/shared_blockly_function.rb (hunk 4 above)
Comment (balderdash) [reply dashboard/app/models/shared_blockly_function.rb]: Opened https://github.com/code-dot-org/code-dot-org/pull/23107 to make the code available as a concern, PTAL at that.

**Pull request merged**